File: backend/src/services/mapillary.ts
import { APP_CONSTANTS } from '../config/env.js';
import { redis } from '../db/redis.js';
import { createLogger } from '../logger/index.js';
import { serviceUnavailable, tooManyRequests } from '../utils/httpError.js';

const log = createLogger('mapillary');

const GRAPH_BASE_URL = 'https://graph.mapillary.com';
const MEDIA_FIELDS = 'thumb_256_url,thumb_1024_url,thumb_2048_url';
const IMAGE_SEARCH_FIELDS = 'id,geometry,is_pano';
const CACHE_CONTENT_TYPE = 'image/jpeg';

const SEARCH_CACHE_PREFIX = 'mly:search:';
const MEDIA_CACHE_PREFIX = 'mly:media:';
const IMAGE_CACHE_PREFIX = 'mly:img:';

// bbox 格式：minLng,minLat,maxLng,maxLat（十进制小数，允许负值）
export const BBOX_PATTERN = /^-?\d+(\.\d+)?,-?\d+(\.\d+)?,-?\d+(\.\d+)?,-?\d+(\.\d+)?$/;

export interface MapillaryImage {
    id: string;
    geometry: { type: 'Point'; coordinates: [number, number] };
    is_pano?: boolean;
}

export interface MapillarySearchResult {
    data: MapillaryImage[];
}

interface MediaRecord {
    thumb_256_url?: string;
    thumb_1024_url?: string;
    thumb_2048_url?: string;
}

// 请求时惰性读取密钥：便于测试注入，且服务启动后更新环境变量无需重启即可生效
function getMapillaryToken(): string {
    return process.env.MAPILLARY_TOKEN ?? '';
}

function assertTokenConfigured(): void {
    if (getMapillaryToken() === '') {
        throw serviceUnavailable('Mapillary 代理未配置（缺少 MAPILLARY_TOKEN）');
    }
}

function mapUpstreamError(status: number): Error {
    if (status === 429) {
        return tooManyRequests('Mapillary 请求超出额度，请稍后再试');
    }
    if (status === 401 || status === 403) {
        log.error(`Mapillary 授权失败（HTTP ${status}），请检查 MAPILLARY_TOKEN`);
        return serviceUnavailable('Mapillary 授权失败，请检查服务端配置');
    }
    return serviceUnavailable(`Mapillary 服务暂不可用（HTTP ${status}）`);
}

async function requestJson(url: string): Promise<unknown> {
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), APP_CONSTANTS.MAPILLARY_TIMEOUT_MS);
    try {
        const response = await fetch(url, {
            signal: controller.signal,
            headers: { 'User-Agent': 'mma-guessr-backend' },
        });
        if (!response.ok) {
            throw mapUpstreamError(response.status);
        }
        return (await response.json()) as unknown;
    } catch (err) {
        if (err instanceof Error && err.name === 'AbortError') {
            throw serviceUnavailable('Mapillary 请求超时');
        }
        throw err;
    } finally {
        clearTimeout(timeoutId);
    }
}

export async function searchMapillaryImages(bbox: string, limit: number): Promise<MapillarySearchResult> {
    assertTokenConfigured();
    const cacheKey = `${SEARCH_CACHE_PREFIX}${bbox}:${limit}`;
    const cached = await redis.get(cacheKey);
    if (cached !== null) {
        return JSON.parse(cached) as MapillarySearchResult;
    }

    const params = new URLSearchParams({
        access_token: getMapillaryToken(),
        fields: IMAGE_SEARCH_FIELDS,
        bbox,
        limit: String(limit),
    });
    const result = (await requestJson(`${GRAPH_BASE_URL}/images?${params.toString()}`)) as MapillarySearchResult;
    await redis.set(cacheKey, JSON.stringify(result), 'EX', APP_CONSTANTS.MAPILLARY_SEARCH_TTL_SECONDS);
    return result;
}

async function resolveMedia(imageId: string): Promise<MediaRecord> {
    const cacheKey = `${MEDIA_CACHE_PREFIX}${imageId}`;
    const cached = await redis.get(cacheKey);
    if (cached !== null) {
        return JSON.parse(cached) as MediaRecord;
    }

    const params = new URLSearchParams({ access_token: getMapillaryToken(), fields: MEDIA_FIELDS });
    const media = (await requestJson(`${GRAPH_BASE_URL}/${imageId}?${params.toString()}`)) as MediaRecord;
    await redis.set(cacheKey, JSON.stringify(media), 'EX', APP_CONSTANTS.MAPILLARY_MEDIA_TTL_SECONDS);
    return media;
}

// 选择不小于请求宽度、最近的一档缩略图；没有则取最大一档
function pickThumbUrl(media: MediaRecord, width: number): string {
    const sizeCandidates = [
        { size: 256, url: media.thumb_256_url },
        { size: 1024, url: media.thumb_1024_url },
        { size: 2048, url: media.thumb_2048_url },
    ].filter((candidate): candidate is { size: number; url: string } => candidate.url !== undefined);

    const chosen =
        sizeCandidates.find((candidate) => candidate.size >= width) ?? sizeCandidates[sizeCandidates.length - 1];
    if (chosen === undefined) {
        throw serviceUnavailable('Mapillary 图片缺少缩略图字段');
    }
    return chosen.url;
}

// 上游返回的缩略图 URL 做基础 SSRF 防护：仅允许 HTTPS，并拒绝回环/内网/链路本地/云元数据地址。
// 不做严格域名白名单，避免 Mapillary CDN 域名变动导致街景加载被误伤；此处聚焦阻断 SSRF 高危目标。
function assertSafeImageUrl(rawUrl: string): string {
    let parsed: URL;
    try {
        parsed = new URL(rawUrl);
    } catch {
        throw serviceUnavailable('图源返回了非法图片地址');
    }
    if (parsed.protocol !== 'https:') {
        throw serviceUnavailable('图源图片地址必须使用 HTTPS');
    }
    const host = parsed.hostname.toLowerCase();
    if (host === '' || host === 'localhost' || host === '0.0.0.0' || host === '127.0.0.1' || host === '::1') {
        throw serviceUnavailable('图源图片地址指向了受保护地址');
    }
    const ipv4 = /^(\d{1,3})\.(\d{1,3})\.(\d{1,3})\.(\d{1,3})$/.exec(host);
    if (ipv4 !== null) {
        const a = Number(ipv4[1]);
        const b = Number(ipv4[2]);
        const isPrivate =
            a === 10 || // 10.0.0.0/8
            a === 0 || // 0.0.0.0/8
            a === 127 || // 127.0.0.0/8
            (a === 172 && b >= 16 && b <= 31) || // 172.16.0.0/12
            (a === 192 && b === 168) || // 192.168.0.0/16
            (a === 169 && b === 254) || // 169.254.0.0/16（链路本地 / 云元数据）
            (a === 100 && b >= 64 && b <= 127); // 100.64.0.0/10（CGNAT）
        if (isPrivate) {
            throw serviceUnavailable('图源图片地址指向了内网地址');
        }
    }
    // IPv6：拒绝回环/未指定/链路本地/唯一本地地址
    if (host === '::' || host.startsWith('fe80:') || host.startsWith('fc') || host.startsWith('fd')) {
        throw serviceUnavailable('图源图片地址指向了内网地址');
    }
    return parsed.toString();
}

async function fetchImageBuffer(url: string): Promise<Buffer> {
    const safeUrl = assertSafeImageUrl(url);
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), APP_CONSTANTS.MAPILLARY_TIMEOUT_MS);
    try {
        const response = await fetch(safeUrl, {
            signal: controller.signal,
            headers: { 'User-Agent': 'mma-guessr-backend' },
        });
        if (!response.ok) {
            throw serviceUnavailable(`Mapillary 图片加载失败（HTTP ${response.status}）`);
        }
        // 单次最大缓存上限同时作为硬性拉取上限：避免上游异常时流式灌入超大内存
        const reader = response.body?.getReader();
        if (reader === undefined) {
            throw serviceUnavailable('Mapillary 图片响应缺少内容体');
        }
        const chunks: Uint8Array[] = [];
        let totalBytes = 0;
        while (true) {
            const { done, value } = await reader.read();
            if (done) {
                break;
            }
            totalBytes += value.byteLength;
            if (totalBytes > APP_CONSTANTS.MAPILLARY_MAX_IMAGE_BYTES) {
                void reader.cancel().catch(() => undefined);
                throw serviceUnavailable('Mapillary 图片超过大小限制');
            }
            chunks.push(value);
        }
        return Buffer.concat(chunks);
    } catch (err) {
        if (err instanceof Error && err.name === 'AbortError') {
            throw serviceUnavailable('Mapillary 图片加载超时');
        }
        throw err;
    } finally {
        clearTimeout(timeoutId);
    }
}

// 宽度归一化到三档（与上游缩略图档位一致）：防止任意 width 膨胀 Redis 缓存条目
const ALLOWED_WIDTHS = [256, 1024, 2048] as const;
// noUncheckedIndexedAccess 下元组索引含 undefined，兜底直接取最大档位字面量
const LARGEST_ALLOWED_WIDTH = 2048;
function normalizeWidth(width: number): number {
    for (const candidate of ALLOWED_WIDTHS) {
        if (candidate >= width) {
            return candidate;
        }
    }
    return LARGEST_ALLOWED_WIDTH;
}

export interface FetchedImage {
    buffer: Buffer;
    contentType: string;
}

export async function fetchMapillaryImage(imageId: string, width: number): Promise<FetchedImage> {
    assertTokenConfigured();
    const normalizedWidth = normalizeWidth(width);
    const media = await resolveMedia(imageId);
    const thumbUrl = pickThumbUrl(media, normalizedWidth);

    const cacheKey = `${IMAGE_CACHE_PREFIX}${imageId}:${normalizedWidth}`;
    const cached = await redis.getBuffer(cacheKey);
    if (cached !== null) {
        return { buffer: cached, contentType: CACHE_CONTENT_TYPE };
    }

    const buffer = await fetchImageBuffer(thumbUrl);
    if (buffer.length <= APP_CONSTANTS.MAPILLARY_MAX_IMAGE_BYTES) {
        await redis.set(cacheKey, buffer, 'EX', APP_CONSTANTS.MAPILLARY_IMAGE_TTL_SECONDS);
    }
    return { buffer, contentType: CACHE_CONTENT_TYPE };
}

File: backend/internal/mapillary/mapillary.go
package mapillary

import (
	"encoding/json"
	"fmt"
	"io"
	"net"
	"net/http"
	"net/url"
	"strings"
	"time"

	"mma-guessr/backend/internal/httputil"
	"mma-guessr/backend/internal/kv"
)

const (
	graphBaseURL     = "https://graph.mapillary.com"
	mediaFields      = "thumb_256_url,thumb_1024_url,thumb_2048_url"
	searchFields     = "id,geometry,is_pano,thumb_256_url,thumb_1024_url,thumb_2048_url"
	imageContentType = "image/jpeg"

	searchCachePrefix = "mly:search:"
	mediaCachePrefix  = "mly:media:"
	imageCachePrefix  = "mly:img:"

	allowedWidths  = 256
	largestWidth   = 2048
	maxImageBytes  = 1 << 20
	defaultTimeout = 10 * time.Second
)

// BBoxPattern describes minLng,minLat,maxLng,maxLat (decimal, negatives ok).

// SearchResult mirrors the Mapillary /images response.
type SearchResult struct {
	Data []Image `json:"data"`
}

// Image is one Mapillary image entry.
type Image struct {
	ID           string    `json:"id"`
	Thumb256URL  *string   `json:"thumb_256_url,omitempty"`
	Thumb1024URL *string   `json:"thumb_1024_url,omitempty"`
	Thumb2048URL *string   `json:"thumb_2048_url,omitempty"`
	Geometry     struct {
		Type        string    `json:"type"`
		Coordinates []float64 `json:"coordinates"`
	} `json:"geometry"`
	IsPano *bool `json:"is_pano,omitempty"`
}

// FetchedImage is the resolved thumbnail bytes.
type FetchedImage struct {
	Buffer      []byte
	ContentType string
}

// Service proxies Mapillary search and imagery with server-held secrets.
type Service struct {
	token        string
	kv           *kv.Store
	client       *http.Client
	baseURL      string
	imageFetcher func(string) ([]byte, error)
}

// NewService creates a Mapillary proxy service.
func NewService(token string, cache *kv.Store) *Service {
	return &Service{
		token:   token,
		kv:      cache,
		client:  &http.Client{Timeout: defaultTimeout},
		baseURL: graphBaseURL,
	}
}

// SetUpstream overrides the upstream base URL and client (tests only).
func (s *Service) SetUpstream(baseURL string, client *http.Client) {
	if baseURL != "" {
		s.baseURL = baseURL
	}
	if client != nil {
		s.client = client
	}
}

// SetToken overrides the Mapillary access token (tests only).
func (s *Service) SetToken(token string) {
	s.token = token
}

// SetImageFetcher overrides the downstream image fetch (tests only). The
// SSRF guard still runs before the fetcher is invoked.
func (s *Service) SetImageFetcher(fetcher func(string) ([]byte, error)) {
	s.imageFetcher = fetcher
}

// IsValidBBox reports whether the bbox string matches the expected format.
func IsValidBBox(bbox string) bool {
	parts := strings.Split(bbox, ",")
	if len(parts) != 4 {
		return false
	}
	for _, part := range parts {
		if !isDecimal(part) {
			return false
		}
	}
	return true
}

func isDecimal(value string) bool {
	if value == "" {
		return false
	}
	hasDot := false
	for i := 0; i < len(value); i++ {
		c := value[i]
		if c == '-' {
			if i != 0 {
				return false
			}
			continue
		}
		if c == '.' {
			if hasDot {
				return false
			}
			hasDot = true
			continue
		}
		if c < '0' || c > '9' {
			return false
		}
	}
	return !hasDot || value != "."
}

// SearchImages queries the Mapillary image search with a cached result.
func (s *Service) SearchImages(bbox string, limit int) (*SearchResult, error) {
	if s.token == "" {
		return nil, httputil.ServiceUnavailable("Mapillary 代理未配置（缺少 MAPILLARY_TOKEN）")
	}
	cacheKey := searchCachePrefix + bbox + ":" + fmt.Sprint(limit)
	if cached, ok := s.kv.Get(cacheKey); ok {
		var result SearchResult
		if err := json.Unmarshal([]byte(cached), &result); err == nil {
			return &result, nil
		}
	}

	params := url.Values{}
	params.Set("access_token", s.token)
	params.Set("fields", searchFields)
	params.Set("bbox", bbox)
	params.Set("limit", fmt.Sprint(limit))
	raw, err := s.requestJSON(s.baseURL + "/images?" + params.Encode())
	if err != nil {
		return nil, err
	}

	var result SearchResult
	if err := json.Unmarshal(raw, &result); err != nil {
		return nil, httputil.ServiceUnavailable("Mapillary 服务暂不可用（响应格式异常）")
	}
	_ = s.kv.Set(cacheKey, string(raw), 24*60*60)
	return &result, nil
}

// FetchImage resolves and proxies a thumbnail, caching the bytes.
func (s *Service) FetchImage(imageID string, width int) (*FetchedImage, error) {
	if s.token == "" {
		return nil, httputil.ServiceUnavailable("Mapillary 代理未配置（缺少 MAPILLARY_TOKEN）")
	}
	normalized := normalizeWidth(width)
	cacheKey := imageCachePrefix + imageID + ":" + fmt.Sprint(normalized)
	if cached, ok := s.kv.GetBytes(cacheKey); ok {
		return &FetchedImage{Buffer: cached, ContentType: imageContentType}, nil
	}

	media, err := s.resolveMedia(imageID)
	if err != nil {
		return nil, err
	}
	thumbURL, err := pickThumbURL(media, normalized)
	if err != nil {
		return nil, err
	}
	buffer, err := s.fetchImage(thumbURL)
	if err != nil {
		return nil, err
	}
	_ = s.kv.SetBytes(cacheKey, buffer, 24*60*60)
	return &FetchedImage{Buffer: buffer, ContentType: imageContentType}, nil
}

// ResolveMediaURL resolves a Mapillary image to its public CDN thumbnail URL.
// Unlike FetchImage it never downloads bytes, so the browser can load the
// image directly from the CDN while the token stays server-side.
func (s *Service) ResolveMediaURL(imageID string, width int) (string, error) {
	if s.token == "" {
		return "", httputil.ServiceUnavailable("Mapillary 代理未配置（缺少 MAPILLARY_TOKEN）")
	}
	media, err := s.resolveMedia(imageID)
	if err != nil {
		return "", err
	}
	thumbURL, err := pickThumbURL(media, normalizeWidth(width))
	if err != nil {
		return "", err
	}
	if err := assertSafeImageURL(thumbURL); err != nil {
		return "", err
	}
	return thumbURL, nil
}

type mediaRecord struct {
	Thumb256URL  *string `json:"thumb_256_url"`
	Thumb1024URL *string `json:"thumb_1024_url"`
	Thumb2048URL *string `json:"thumb_2048_url"`
}

func (s *Service) resolveMedia(imageID string) (*mediaRecord, error) {
	cacheKey := mediaCachePrefix + imageID
	if cached, ok := s.kv.Get(cacheKey); ok {
		var media mediaRecord
		if err := json.Unmarshal([]byte(cached), &media); err == nil {
			return &media, nil
		}
	}
	params := url.Values{}
	params.Set("access_token", s.token)
	params.Set("fields", mediaFields)
	raw, err := s.requestJSON(s.baseURL + "/" + imageID + "?" + params.Encode())
	if err != nil {
		return nil, err
	}
	var media mediaRecord
	if err := json.Unmarshal(raw, &media); err != nil {
		return nil, httputil.ServiceUnavailable("Mapillary 服务暂不可用（响应格式异常）")
	}
	_ = s.kv.Set(cacheKey, string(raw), 24*60*60)
	return &media, nil
}

func (s *Service) requestJSON(target string) ([]byte, error) {
	req, err := http.NewRequest(http.MethodGet, target, nil)
	if err != nil {
		return nil, httputil.ServiceUnavailable("Mapillary 服务暂不可用")
	}
	req.Header.Set("User-Agent", "mma-guessr-backend")
	response, err := s.client.Do(req)
	if err != nil {
		return nil, httputil.ServiceUnavailable("Mapillary 请求超时")
	}
	defer response.Body.Close()
	if response.StatusCode != http.StatusOK {
		return nil, mapUpstreamError(response.StatusCode)
	}
	raw, err := io.ReadAll(io.LimitReader(response.Body, 4<<20))
	if err != nil {
		return nil, httputil.ServiceUnavailable("Mapillary 服务暂不可用")
	}
	return raw, nil
}

func mapUpstreamError(status int) error {
	switch status {
	case http.StatusTooManyRequests:
		return httputil.TooManyRequests("Mapillary 请求超出额度，请稍后再试")
	case http.StatusUnauthorized, http.StatusForbidden:
		return httputil.ServiceUnavailable("Mapillary 授权失败，请检查服务端配置")
	default:
		return httputil.ServiceUnavailable(fmt.Sprintf("Mapillary 服务暂不可用（HTTP %d）", status))
	}
}

func pickThumbURL(media *mediaRecord, width int) (string, error) {
	candidates := []struct {
		size int
		url  *string
	}{
		{256, media.Thumb256URL},
		{1024, media.Thumb1024URL},
		{2048, media.Thumb2048URL},
	}
	var chosen *string
	for _, candidate := range candidates {
		if candidate.url != nil {
			if candidate.size >= width {
				chosen = candidate.url
				break
			}
			chosen = candidate.url
		}
	}
	if chosen == nil {
		return "", httputil.ServiceUnavailable("Mapillary 图片缺少缩略图字段")
	}
	return *chosen, nil
}

// normalizeWidth snaps to the nearest supported thumbnail tier.
func normalizeWidth(width int) int {
	switch {
	case width <= 256:
		return allowedWidths
	case width <= 1024:
		return 1024
	default:
		return largestWidth
	}
}

// assertSafeImageURL blocks SSRF targets while keeping CDN flexibility.
func assertSafeImageURL(rawURL string) error {
	parsed, err := url.Parse(rawURL)
	if err != nil || parsed.Host == "" {
		return httputil.ServiceUnavailable("图源返回了非法图片地址")
	}
	if parsed.Scheme != "https" {
		return httputil.ServiceUnavailable("图源图片地址必须使用 HTTPS")
	}
	host := strings.ToLower(parsed.Hostname())
	if host == "" || host == "localhost" || host == "0.0.0.0" || host == "127.0.0.1" || host == "::1" {
		return httputil.ServiceUnavailable("图源图片地址指向了受保护地址")
	}
	if ip := net.ParseIP(host); ip != nil {
		if ip.IsLoopback() || ip.IsPrivate() || ip.IsLinkLocalUnicast() || ip.IsLinkLocalMulticast() ||
			ip.IsUnspecified() || ip.IsMulticast() {
			return httputil.ServiceUnavailable("图源图片地址指向了内网地址")
		}
		// CGNAT 100.64.0.0/10 and the 0.0.0.0/8 legacy range.
		if ip.To4() != nil {
			octets := ip.To4()
			if octets[0] == 100 && octets[1] >= 64 && octets[1] <= 127 {
				return httputil.ServiceUnavailable("图源图片地址指向了内网地址")
			}
			if octets[0] == 0 || octets[0] == 169 && octets[1] == 254 {
				return httputil.ServiceUnavailable("图源图片地址指向了内网地址")
			}
		}
	}
	return nil
}

func (s *Service) fetchImage(rawURL string) ([]byte, error) {
	if err := assertSafeImageURL(rawURL); err != nil {
		return nil, err
	}
	if s.imageFetcher != nil {
		return s.imageFetcher(rawURL)
	}
	return s.fetchImageBuffer(rawURL)
}

func (s *Service) fetchImageBuffer(rawURL string) ([]byte, error) {
	req, err := http.NewRequest(http.MethodGet, rawURL, nil)
	if err != nil {
		return nil, httputil.ServiceUnavailable("Mapillary 图片加载失败")
	}
	req.Header.Set("User-Agent", "mma-guessr-backend")
	response, err := s.client.Do(req)
	if err != nil {
		return nil, httputil.ServiceUnavailable("Mapillary 图片加载超时")
	}
	defer response.Body.Close()
	if response.StatusCode != http.StatusOK {
		return nil, httputil.ServiceUnavailable(fmt.Sprintf("Mapillary 图片加载失败（HTTP %d）", response.StatusCode))
	}
	buffer, err := io.ReadAll(io.LimitReader(response.Body, maxImageBytes+1))
	if err != nil {
		return nil, httputil.ServiceUnavailable("Mapillary 图片加载失败")
	}
	if len(buffer) > maxImageBytes {
		return nil, httputil.ServiceUnavailable("Mapillary 图片超过大小限制")
	}
	return buffer, nil
}
